server/util/db/db.go

```diff
@@ -119,12 +121,80 @@ func maybeRunMigrations(dialect string, gdb *gorm.DB) error {
 	return nil
 }
 
-func openDB(dialect string, args ...interface{}) (*gorm.DB, error) {
-	gdb, err := gorm.Open(dialect, args...)
+// instrumentGORM adds GORM callbacks that populate query metrics.
+func instrumentGORM(gdb *gorm.DB) {
+	// Add callback that runs before other callbacks that records when the operation began.
+	// We use this to calculate how long a query takes to run.
+	recordStartTime := func(db *gorm.DB) {
+		if db.DryRun || db.Statement == nil {
+			return
+		}
+		db.Statement.Settings.Store(gormStmtStartTimeKey, time.Now())
+	}
+	gdb.Callback().Create().Before("*").Register("buildbuddy:record_op_start_time", recordStartTime)
+	gdb.Callback().Delete().Before("*").Register("buildbuddy:record_op_start_time", recordStartTime)
+	gdb.Callback().Query().Before("*").Register("buildbuddy:record_op_start_time", recordStartTime)
+	gdb.Callback().Raw().Before("*").Register("buildbuddy:record_op_start_time", recordStartTime)
+	gdb.Callback().Row().Before("*").Register("buildbuddy:record_op_start_time", recordStartTime)
+	gdb.Callback().Update().Before("*").Register("buildbuddy:record_op_start_time", recordStartTime)
+
+	// Add callback that runs after other callbacks that records executed queries and their durations.
+	recordMetrics := func(db *gorm.DB) {
+		if db.DryRun || db.Statement == nil {
+			return
+		}
+		labels := prometheus.Labels{
+			metrics.SQLQueryTemplateLabel: db.Statement.SQL.String(),
+		}
+		v, _ := db.Statement.Settings.LoadAndDelete(gormStmtStartTimeKey)
```

Comment:
If this code ignores the error returned by LoadAndDelete -- can you add a short comment about why this is OK? Just is sort of a code-smell to ignore errors like this so when we do it I think we should be explicit about why!

(seems like if there is an err, v is bad, then the query duration won't be logged which is fine! and we still will count the sqlerrorcount, which is what we want)

Reply:
Settings is a sync.Map so the second value indicates whether the key was present in the map.
I'm relying on checking whether v is nil to do the same. I've added a comment to clarify.